### PR TITLE
feat: update test to correctly expect fork promotion

### LINKED ISSUE #3: Chainwork is not computed correctly on neither mainnet, nor forks

### PR DESCRIPTION
**Pull request by raventid**
Fixes #3 . Besides also include minor changes to public contract interfaces. This is a WIP PR, but you can take a look at unit tests and interfaces and ask questions here!

### CODE REVIEW

File: contract/src/lib.rs [hunk 1]
```diff
@@ -237,70 +288,112 @@ impl Contract {
...
+    pub fn receive_state(&self) -> std::collections::BTreeMap<String, usize> {
+        let mut state = std::collections::BTreeMap::new();
 
         // Add last mainnet block to the state
-        state.insert(0, self.heaviest_block.clone());
+        state.insert(self.heaviest_block.clone(), 0);
 
-        // Extract all of the last fork blocks
-        for fork_id in 1 ..= self.current_fork_id {
+        for fork_id in self.alive_forks.iter() {
```
Comment (raventid) [inline contract/src/lib.rs:358]: iter() is not good for a smart contract, maybe remove iteration?

File: contract/src/lib.rs [hunk 2]
```diff
@@ -73,56 +73,61 @@ mod state {
...
-    forks: near_sdk::store::LookupMap<usize, Vec<state::Header>>,
-
-    // fork_id -> latest_header
-    latest_fork_block_hashes: near_sdk::store::LookupMap<usize, state::Header>,
-
-    // mapping(bytes32 => HeaderInfo) public _headers;
-    // mapping(uint256 => bytes32) public _mainChain; // mapping of block heights to block hashes of the MAIN CHAIN
+    forks: near_sdk::store::LookupMap<usize, ForkState>,
+    // still alive forks, others would be garbage collected
+    alive_forks: std::collections::HashSet<usize>,
```
Comment (raventid) [inline contract/src/lib.rs:102]: Do not use HashSet for storage

File: contract/src/lib.rs (hunk 1 above)
Comment (raventid) [reply contract/src/lib.rs]: algorithm has changed

File: contract/src/lib.rs (hunk 2 above)
Comment (raventid) [reply contract/src/lib.rs]: removed

File: contract/src/lib.rs [hunk 3]
```diff
@@ -73,59 +87,45 @@ mod state {
...
     fn default() -> Self {
         Self {
-            height_to_header: near_sdk::store::LookupMap::new(b"a"),
-            headers: near_sdk::store::LookupMap::new(b"h"),
-            forks: near_sdk::store::LookupMap::new(b"f"),
-            latest_fork_block_hashes: near_sdk::store::LookupMap::new(b"l"),
-            current_fork_id: 0,
-            heaviest_block: String::new(),
-            high_score: [0; 32],
+            mainchain_height_to_header: near_sdk::store::LookupMap::new(b"a"),
```
Comment (karim-en) [inline contract/src/lib.rs:125]: Please use enum:
#[derive(BorshSerialize, BorshStorageKey)]
enum StorageKey {
    MainchainHeightToHeader,
    MainchainHeaderToHeight,
    ....
}

Comment (raventid) [reply contract/src/lib.rs]: Done

File: contract/src/lib.rs [hunk 4]
```diff
@@ -73,258 +85,320 @@ mod state {
...
     fn default() -> Self {
         Self {
-            height_to_header: near_sdk::store::LookupMap::new(b"a"),
-            headers: near_sdk::store::LookupMap::new(b"h"),
-            forks: near_sdk::store::LookupMap::new(b"f"),
-            latest_fork_block_hashes: near_sdk::store::LookupMap::new(b"l"),
-            current_fork_id: 0,
-            heaviest_block: String::new(),
-            high_score: [0; 32],
+            mainchain_height_to_header: near_sdk::store::LookupMap::new(
```
Comment (karim-en) [inline contract/src/lib.rs:109]: Let's remove the Default crate implementation, and move the this code to the init method

File: contract/src/lib.rs [hunk 5]
```diff
@@ -73,258 +85,320 @@ mod state {
...
+        contract.submit_genesis(genesis_block, genesis_block_height);
+
+        contract
     }
 
-    // TODO: Should we submit genesis block separately or we can try to find the way to do it as a part
-    // TODO: of a general flow?
-    pub fn submit_genesis(&mut self, block_header: Header) -> bool {
-        let current_block_hash= block_header.block_hash().as_raw_hash().to_string();
+    fn submit_genesis(&mut self, block_header: Header, block_height: u64) -> bool {
```
Comment (karim-en) [inline contract/src/lib.rs:139]: Let's rename to init_genesis

File: contract/src/lib.rs (hunk 4 above)
Comment (raventid) [reply contract/src/lib.rs]: done

File: contract/src/lib.rs (hunk 5 above)
Comment (raventid) [reply contract/src/lib.rs]: done

**Pull request merged**